flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobSubmitHandler.java

```diff
@@ -54,18 +69,89 @@ public JobSubmitHandler(
+		for (String jarFileName : requestBody.jarFileNames) {
+			Path jarFile = getPathAndAssertUpload(jarFileName, "Jar", nameToFile);
+			jarFiles.add(new org.apache.flink.core.fs.Path(jarFile.toString()));
+		}
+
+		Collection<Tuple2<String, org.apache.flink.core.fs.Path>> artifacts = new ArrayList<>(requestBody.artifactFileNames.size());
+		for (JobSubmitRequestBody.DistributedCacheFile artifactFileName : requestBody.artifactFileNames) {
+			Path artifactFile = getPathAndAssertUpload(artifactFileName.fileName, "Artifact", nameToFile);
+			artifacts.add(Tuple2.of(artifactFileName.entryName, new org.apache.flink.core.fs.Path(artifactFile.toString())));
 		}
 
-		return gateway.submitJob(jobGraph, timeout)
-			.thenApply(ack -> new JobSubmitResponseBody("/jobs/" + jobGraph.getJobID()));
+		Map<String, DistributedCache.DistributedCacheEntry> temporaryHack = artifacts.stream()
+			.collect(Collectors.toMap(
+				tuple -> tuple.f0,
+				// the actual entry definition is mostly irrelevant as only the blobkey is accessed
+				// blame whoever wrote the ClientUtils API
+				tuple -> new DistributedCache.DistributedCacheEntry(tuple.f1.toString(), false)
+			));
+
+		// TODO: use executor
+		CompletableFuture<JobGraph> jobGraphFuture = CompletableFuture.supplyAsync(() -> {
+			JobGraph jobGraph;
+			try (ObjectInputStream objectIn = new ObjectInputStream(Files.newInputStream(jobGraphFile))) {
+				jobGraph = (JobGraph) objectIn.readObject();
+			} catch (Exception e) {
+				throw new CompletionException(new RestHandlerException(
+					"Failed to deserialize JobGraph.",
+					HttpResponseStatus.BAD_REQUEST,
+					e));
+			}
+			return jobGraph;
+		});
+
+		CompletableFuture<Integer> blobServerPortFuture = gateway.getBlobServerPort(timeout);
+
+		CompletableFuture<JobGraph> finalizedJobGraphFuture = jobGraphFuture.thenCombine(blobServerPortFuture, (jobGraph, blobServerPort) -> {
+			final InetSocketAddress address = new InetSocketAddress(gateway.getHostname(), blobServerPort);
+			try (BlobClient blobClient = new BlobClient(address, new Configuration())) {
+				Collection<PermanentBlobKey> jarBlobKeys = ClientUtils.uploadUserJars(jobGraph.getJobID(), jarFiles, blobClient);
+				ClientUtils.setUserJarBlobKeys(jarBlobKeys, jobGraph);
+
+				Collection<Tuple2<String, PermanentBlobKey>> artifactBlobKeys = ClientUtils.uploadUserArtifacts(jobGraph.getJobID(), temporaryHack, blobClient);
+				ClientUtils.setUserArtifactBlobKeys(jobGraph, artifactBlobKeys);
+			} catch (IOException e) {
+				throw new CompletionException(new RestHandlerException(
+					"Could not upload job files.",
+					HttpResponseStatus.INTERNAL_SERVER_ERROR,
+					e));
+			}
+			return jobGraph;
+		});
+
+		CompletableFuture<Acknowledge> jobSubmissionFuture = finalizedJobGraphFuture.thenCompose(jobGraph -> gateway.submitJob(jobGraph, timeout));
+
+		return jobSubmissionFuture.thenCombine(jobGraphFuture,
+			(ack, jobGraph) -> new JobSubmitResponseBody("/jobs/" + jobGraph.getJobID()));
+	}
```

Comment:
This method is quite long. We could think about moving the generation of the `jarFiles` and `temporaryHack` structures into individual methods.

Reply:
Could we maybe further split this method up into `loadJobGraph`, `uploadJobGraphFiles` and `submitJobGraph`?

Reply:
Nice, it's now much easier to understand 👍